### ISSUE TITLE
Service.go is part of the application layer on the hexagonal architecture

### ISSUE THREAD

**Issue by K4L1Ma** [kasperbe/port-processor#3]
Service.go does the orchestration of decoding the json into a port so then we can Upsert it into the db, as a piece orchestrating different layers of the hexagonal it natural fit would be the application layer and not the domain layer

**Comment by K4L1Ma** [comment]
I really like the stream approach would have love to see a better usage, like for example, having an in-memory sync.Map would have allowed you to parallelize the upserting of the ports allowing you to even config the number of file consumers given that the file could be huge in size

**Comment by K4L1Ma** [comment]
also context handling on Import

**Comment by kasperbe** [comment]
> Service.go do the orchestration of decoding the JSON (i/o) into a port(domain) so then we can Upsert it into the DB(i/o), as a piece orchestrating different layers of the hexagonal its natural fit would be the application layer and not the domain layer

You're absolutely right. I actually did have it split out, but for some reason decided to mash it back into the domain service. I think I ended up mashing too many things together just to keep it simpler, and breaking hexagonal pattern in doing so. 

---


> I really like the stream approach and would love to see a better usage, like for example, having an in-memory sync.Map would have allowed you to parallelize the upserting of the ports allowing you to even config the number of file consumers given that the file could be huge in size

This is a good point, however I thought from reading through the requirements, that ports could be in the file multiple times, and the newest should win. If I parallelized the import step, it would become a little tricky to figure out the final state of the data, since there's no other indicator on each entry to indicate whether it's newer or older.

```
The end result should be a storage containing the ports, representing the latest version found in the JSON.
```

**Comment by K4L1Ma** [comment]
probably going out of scope but possibly having an optimistic lock based on the read/consumed time, so you never override the latest version found on the JSON would be more than enough to accomplish such a feature, as you read sequentially doesn't mean you need to consume/upsert in the same order

**Comment by kasperbe** [comment]
Aah, I see what you mean, not parrallelizing the importer, but the upsering. Sorry, I misread your initial comment. You're right, that's absolutely possible :)
